src/nccl_ofi_api.c

```diff
@@ -308,57 +315,139 @@ ncclResult_t nccl_net_ofi_regMr(void *comm, void *data, size_t size, int type,
 		return ncclInternalError;
 	}
 
-	switch (base_comm->type) {
-	case NCCL_NET_OFI_SEND_COMM:;
-		nccl_net_ofi_send_comm_t *send_comm =
-			(nccl_net_ofi_send_comm_t *)base_comm;
-		ret = send_comm->regMr(send_comm, data, size, type, mhandle);
-		break;
-	case NCCL_NET_OFI_RECV_COMM:;
-		nccl_net_ofi_recv_comm_t *recv_comm =
-			(nccl_net_ofi_recv_comm_t *)base_comm;
-		ret = recv_comm->regMr(recv_comm, data, size, type, mhandle);
-		break;
-	default:
-		NCCL_OFI_WARN("Unexpected communicator type. Communicator type: %d",
-			      base_comm->type);
-		ret = -EINVAL;
-		break;
+	page_size = (uintptr_t) system_page_size;
+	addr = (uintptr_t)data & -page_size; /* start of page of data */
+	pages = ((uintptr_t)data + size - addr + page_size-1)/page_size; /* Number of pages in buffer */
+	dev = base_comm->ep->device;
+	cache = &dev->mr_cache;
+	assert(dev);
+	assert(cache);
+
+	/* NCCL_REG is a better subsys to log these to, but ext-net headers do not have the updated enums yet. */
+	NCCL_OFI_TRACE(NCCL_NET, "Reg addr %lx data %p pages %zu pgsz %zu", addr, data, pages, page_size);
+
+	nccl_net_ofi_mutex_lock(&cache->lock);
+	for (int slot = 0;; slot++) {
+		if (slot == cache->used || addr < cache->slots[slot]->addr) { /* Cache miss */
+			if (cache->used == cache->size) { /* Need to grow cache */
+				cache->size = cache->size < NCCL_OFI_MR_CACHE_SIZE ?
+					      NCCL_OFI_MR_CACHE_SIZE : cache->size * 2;
+				NCCL_OFI_TRACE(NCCL_NET, "Growing cache to size %d", cache->size);
+				ptr =  realloc(cache->slots, cache->size * sizeof(*cache->slots));
+				if (!ptr) {
+					NCCL_OFI_WARN("Unable to grow cache");
+					goto out;
+				}
+				cache->slots = ptr;
+			}
+			assert(cache->slots);
+			memmove(cache->slots+slot+1, cache->slots+slot, (cache->size - slot) * sizeof(nccl_ofi_reg_entry_t*));
```

Comment:
If I understand this correctly, if `slot == cache->used == cache->size -1` then `cache->slots+slot+1` is overflowing, right? I am also wondering why we need this at all.

Reply:
Yes, it would be overflowing, except that cache was grown during [the previous if](https://github.com/aws/aws-ofi-nccl/pull/447/files/cf7be8e01b47a0c2e701f41ea31ad74af8a3373c#diff-2acb5fac4fc89215d450fe2b45533a9c5d85c8da7a2a6a3b03314e22e8d67da8R332).

As for why we need it, hopefully the previous comment explained that. 😄

Reply:
In my example 
`slot == cache->used == cache->size - 1`
while the previous if has `cache->used == cache->size` so it would be false

Reply:
`cache->size` is the capacity of the cache -- the number of slots that have been allocated. `cache->used` is the number of slots that are occupied by cache entries. So it should never be the case that `cache->used` is larger than `cache->size` -- that means that there are more slots occupied than the total capacity.

The cache growing logic should prevent this situation.

Reply:
Looking again, I misunderstood Amedeo's question. I think he's right: in the case he pointed out, the call to memmove will overflow.

In this call to `memmove`, the third argument should use `cache->used - slot` instead of `cache->size - slot`. That's what NCCL does ([link](https://github.com/NVIDIA/nccl/blob/178b6b759074597777ce13438efb0e0ba625e429/src/register.cc#L119); where their "population" is what we call "used") -- I think we just typo'd when translating.